src/components/Card/Card.test.js

```diff
@@ -0,0 +1,117 @@
+  useHistory: () => ({
+    push: mockUseHistory,
+  }),
+}));
+
+const props = {
+  card: { id: 'p1', title: 'Title', text: 'Text', check: false },
+  onSaveCardData: jest.fn(),
+  readOnly: false,
+};
+
+describe('test Card component', () => {
+  const wrapper = shallow(<Card {...props} />);
+
+  test('test Card render', () => {
+    expect(wrapper.find('CardHeader')).toHaveLength(1);
+    expect(wrapper.find('CardBody')).toHaveLength(1);
+  });
+
+  test('test useEffect, should calls useEffect hook', () => {
+    withHooks(() => {
+      const useEffectHook = jest.spyOn(React, 'useEffect');
+      expect(useEffectHook).not.toBeCalled();
+      wrapper.setProps({ readOnly: true });
+      wrapper.find('CardHeader').prop('onEditClick')();
+      expect(useEffectHook).toBeCalled();
+      useEffectHook.mockRestore();
+    });
+  });
+
+  describe('test useState, should change states in all scenarios', () => {
+    const setStateMock = jest.fn();
+    let wrapper;
+
+    beforeAll(() => {
+      jest.spyOn(React, 'useState').mockImplementation((state) => {
+        return [state, setStateMock];
+      });
+
+      wrapper = shallow(<Card {...props} />);
+    });
+
+    afterAll(() => {
+      jest.restoreAllMocks();
+    });
+
+    test('test onTitleChange', () => {
+      const titleValue = 'Changed title';
+      wrapper.find('CardHeader').prop('onTitleChange')(titleValue);
+      expect(setStateMock).toBeCalledWith({
+        enteredText: props.card.text,
+        enteredTitle: titleValue,
+      });
+    });
+
+    test('test onTextChange', () => {
+      const textValue = 'Changed text';
+      wrapper.find('CardBody').prop('onTextChange')(textValue);
+      expect(setStateMock).toBeCalledWith({
+        enteredTitle: props.card.title,
+        enteredText: textValue,
+      });
+    });
+
+    test('test onCancelClick', () => {
+      wrapper.find('CardHeader').prop('onCancelClick')();
+      expect(setStateMock).toBeCalledWith({
+        enteredText: props.card.text,
+        enteredTitle: props.card.title,
+      });
+      expect(setStateMock).toBeCalledWith(false);
+    });
+  });
+
+  describe('test onSaveCardData, should call onSaveCardData in all scenarios', () => {
+    const wrapper = shallow(<Card {...props} />);
+
+    test('test handleClick', () => {
+      wrapper.find('CardHeader').prop('handleClick')();
+      expect(props.onSaveCardData).toBeCalledWith({
+        id: props.card.id,
+        check: !props.card.check,
+      });
+    });
+
+    test('test onEditClick', () => {
+      wrapper.find('CardHeader').prop('onEditClick')();
+      expect(props.onSaveCardData).toBeCalledWith({
+        id: props.card.id,
+        check: false,
+      });
+    });
+
+    test('test onSave', () => {
+      wrapper.find('CardHeader').prop('onSaveClick')();
+      expect(props.onSaveCardData).toBeCalledWith({
+        id: props.card.id,
+        title: props.card.title,
+        text: props.card.text,
+      });
+    });
```

Comment:
Could you please also check that onSaveCardData is also called with 'manually' changed title and text values?